generative/networks/nets/cdpm.py

```diff
@@ -0,0 +1,589 @@
+        )
+
+        if self.out_channels == channels:
+            self.skip_connection = nn.Identity()
+        elif use_conv:
+            self.skip_connection = Convolution(
+                dims, channels, self.out_channels, padding=1
+            )
+        else:
+            self.skip_connection = Convolution(dims, channels, self.out_channels, kernel_size = 1)
+
+    def forward(self, x: torch.Tensor, emb: torch.Tensor)-> torch.Tensor:
+        """
+        Apply the block to a Tensor, conditioned on a timestep embedding.
+        :param x: an [N x C x ...] Tensor of features.
+        :param emb: an [N x emb_channels] Tensor of timestep embeddings.
+        :return: an [N x C x ...] Tensor of outputs.
+        """
+        h = self.in_layers(x)
+        emb_out = self.emb_layers(emb).type(h.dtype)
+        while len(emb_out.shape) < len(h.shape):
+            emb_out = emb_out[..., None]
+        if self.use_scale_shift_norm:
+            out_norm, out_rest = self.out_layers[0], self.out_layers[1:]
+            scale, shift = torch.chunk(emb_out, 2, dim=1)
+            h = out_norm(h) * (1 + scale) + shift
+            h = out_rest(h)
+        else:
+            h = h + emb_out
+            h = self.out_layers(h)
+        return self.skip_connection(x) + h
+
+
+class FactorizedAttentionBlock(nn.Module):
+
+    def __init__(
+        self, 
+        channels: int, 
+        num_heads: int, 
+        time_embed_dim: int | None=None
+        )-> None:
+        super().__init__()
+        self.spatial_attention = RPEAttention(
+            channels=channels, num_heads=num_heads, time_embed_dim = time_embed_dim, use_rpe_q=False, use_rpe_k=False, use_rpe_v=False,
+        )
+        self.temporal_attention = RPEAttention(
+            channels=channels, num_heads=num_heads,
+            time_embed_dim=time_embed_dim,
+        )
+
+    def forward(self, x: torch.Tensor, attn_mask: torch.Tensor, temb: torch.Tensor, T: int, frame_indices: torch.Tensor|None=None)-> torch.Tensor:
+        BT, C, H, W = x.shape
+        B = BT//T
+        # reshape to have T in the last dimension becuase that's what we attend over
+        x = x.view(B, T, C, H, W).permute(0, 3, 4, 2, 1)  # B, H, W, C, T
+        x = x.reshape(B, H*W, C, T)
+        x = self.temporal_attention(x,
+                                    temb,
+                                    frame_indices,
+                                    attn_mask=attn_mask.flatten(start_dim=2).squeeze(dim=2), # B x T
+                                    )
+
+        # Now we attend over the spatial dimensions by reshaping the input
+        x = x.view(B, H, W, C, T).permute(0, 4, 3, 1, 2)  # B, T, C, H, W
+        x = x.reshape(B, T, C, H*W)
+        x = self.spatial_attention(x,
+                                   temb,
+                                   frame_indices=None,
+                                   )
+        x = x.reshape(BT, C, H, W)
+        return x
+
+
+class UNet_2Plus1_Model(nn.Module):

```

Comment:
would this better be called CDPM?

Reply:
We should also have here docstring description of what this model is and what it's for, explaining the conditional component with possible use examples. Perhaps even `ConditionalDPM` as a name to make it clearer what this is.